plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -285,60 +529,334 @@ public boolean deleteNuageVspDevice(DeleteNuageVspDeviceCmd cmd) {
-                    TimeUnit.MILLISECONDS);
-            scheduler.scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE_NTWK_MACRO"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes,
-                    TimeUnit.MILLISECONDS);
-            scheduler
-                    .scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes, TimeUnit.MILLISECONDS);
-        } else {
-            s_logger.warn("NuageVsp configuration for syncWorkers=" + numOfSyncThreads + " syncInterval=" + syncUpIntervalInMinutes
-                    + " could not be read properly. So, check if the properties are configured properly in global properties");
+            }
+        }
+
+        if (validateDomains) {
+            auditDomainsOnVsp(host, true, false);
         }
     }
 
-    public class NuageVspSyncTask implements Runnable {
+    private boolean auditDomainsOnVsp(HostVO host, boolean add, boolean remove) {
+        List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listByHost(host.getId());
+        if (!CollectionUtils.isEmpty(nuageVspDevices)) {
+            _hostDao.loadDetails(host);
+            List<DomainVO> allDomains = _domainDao.listAll();
+            for (DomainVO domain : allDomains) {
+                SyncDomainCommand cmd = new SyncDomainCommand(domain.getUuid(), domain.getName(), domain.getPath(), add, remove);
+                SyncDomainAnswer answer = (SyncDomainAnswer) _agentMgr.easySend(host.getId(), cmd);
+                return answer.getSuccess();
+            }
+        }
+        return true;
+    }
 
-        private String nuageVspEntity;
+    private String findNuageVspCmsIdForDevice(long deviceId, ConfigurationVO cmsIdConfig) {
+        String configValue = cmsIdConfig.getValue();
+        if (!Strings.isNullOrEmpty(configValue)) {
+            String[] configuredNuageVspDevices = configValue.split(";");
+            for (String configuredNuageVspDevice : configuredNuageVspDevices) {
+                if (configuredNuageVspDevice.startsWith(deviceId + ":")) {
+                    String[] split = configuredNuageVspDevice.split(":");
+                    if (split.length != 2 || (split.length > 1 && Strings.isNullOrEmpty(split[1]))) {
+                        throw new IllegalArgumentException("The configured CMS ID for Nuage VSP device " + deviceId + " is in an incorrect format");
+                    }
+                    return split[1];
+                }
+            }
+        }
+        return null;
+    }
 
-        public NuageVspSyncTask(String nuageVspEntity) {
-            this.nuageVspEntity = nuageVspEntity;
+    public List<String> getDnsDetails(Network network) {
+        List<String> dnsServers = null;
+        Boolean configureDns = Boolean.valueOf(_configDao.getValue(NuageVspManager.NuageVspConfigDns.key()));
+        if (configureDns) {
```

Comment:
To reduce method complexity, short-circuit the method if the ``configureDns`` list is empty by returning an empty list when it is false.